File: python/weftspace/builder.py
# Copyright (c) 2025 by mOctave
#
# This program is free software: you can redistribute it and/or modify it under the
# terms of the GNU Affero General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License along with
# this program. If not, see <https://www.gnu.org/licenses/>.

"""
A class of utility methods designed to allow easy conversion from nodes to objects.
"""

from abc import ABC

from .builder_error import BuilderError
from .data_node import DataNode

class Builder(ABC):
	"""A class of utility methods designed to allow easy conversion from nodes to objects."""

	@classmethod
	def build_string(cls, node: DataNode, arg: int) -> str:
		"""
		Takes an argument from a node and returns it as a string, handling any
		exceptions that occur.
		"""
		try:
			return node.args[arg]
		except IndexError as exc:
			raise BuilderError(f"No argument at position {arg}.", node) from exc



	@classmethod
	def build_int(cls, node: DataNode, arg: int) -> int:
		"""
		Takes an argument from a node and returns it as an integer, handling any
		exceptions that occur.
		"""
		try:
			return int(node.args[arg])
		except ValueError as exc:
			raise BuilderError(
				f"The string \"{node.args[arg]}\" could not be parsed to an integer.", node) from exc
		except IndexError as exc:
			raise BuilderError(f"No argument at position {arg}.", node) from exc



	@classmethod
	def build_float(cls, node: DataNode, arg: int) -> float:
		"""
		Takes an argument from a node and returns it as a float, handling any
		exceptions that occur.
		"""
		try:
			return float(node.args[arg])
		except ValueError as exc:
			raise BuilderError(
				f"The string \"{node.args[arg]}\" could not be parsed to a float.", node) from exc
		except IndexError as exc:
			raise BuilderError(f"No argument at position {arg}.", node) from exc




	@classmethod
	def search(cls, node: DataNode, scope: DataNode) -> DataNode | None:
		"""
		Takes a node and uses it as a key to search the scope for a node
		with a matching name and first argument, handling any exceptions that occur.
		"""
		try:
			for child in scope.children:
				if node.name == child.name and node.args[0] == child.args[0]:
					return child
		except IndexError:
			# This is not necessarily an error, so no error message is printed.
			pass

		return None

File: java/src/main/java/io/github/moctave/weftspace/Builder.java
// Copyright (c) 2025 by mOctave
//
// This program is free software: you can redistribute it and/or modify it under the
// terms of the GNU Affero General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.

package io.github.moctave.weftspace;

import org.jspecify.annotations.*;

/** A class of utility methods designed to allow easy conversion from nodes to objects. */
public abstract class Builder {
	/**
	 * Takes an argument from a node and returns it as a string, throwing a
	 * BuilderException if there is an error that prevents parsing.
	 * 
	 * @param node The node to access.
	 * @param arg The index of the argument to convert.
	 * @return The argument, as a string.
	 */
	public static @NonNull String buildString(@NonNull DataNode node, int arg) throws BuilderException {
		try {
			return node.getArg(arg);
		} catch (IndexOutOfBoundsException e) {
			throw new BuilderException(String.format("No argument at position %d.", arg), node);
		}
	}



	/**
	 * Takes an argument from a node and returns it as an integer, throwing a
	 * BuilderException if there is an error that prevents parsing.
	 * 
	 * @param node The node to access.
	 * @param arg The index of the argument to convert.
	 * @return The argument, as a integer.
	 */
	public static int buildInt(@NonNull DataNode node, int arg) throws BuilderException {
		try {
			return Integer.parseInt(node.getArg(arg));
		} catch (IndexOutOfBoundsException e) {
			throw new BuilderException(String.format("No argument at position %d.", arg), node);
		} catch (NumberFormatException e) {
			throw new BuilderException(
				String.format("The string \"%s\" could not be parsed to an integer.", node.getArg(arg)), node);
		}
	}



	/**
	 * Takes an argument from a node and returns it as a double, throwing a
	 * BuilderException if there is an error that prevents parsing.
	 * 
	 * @param node The node to access.
	 * @param arg The index of the argument to convert.
	 * @return The argument, as a double.
	 */
	public static double buildDouble(@NonNull DataNode node, int arg) throws BuilderException {
		try {
			return Double.parseDouble(node.getArg(arg));
		} catch (IndexOutOfBoundsException e) {
			throw new BuilderException(String.format("No argument at position %d.", arg), node);
		} catch (NumberFormatException e) {
			throw new BuilderException(
				String.format("The string \"%s\" could not be parsed to a double.", node.getArg(arg)), node);
		}
	}



	/**
	 * Takes an argument from a node and returns it as a long int, throwing a
	 * BuilderException if there is an error that prevents parsing.
	 * 
	 * @param node The node to access.
	 * @param arg The index of the argument to convert.
	 * @return The argument, as a long int.
	 */
	public static long buildLong(@NonNull DataNode node, int arg) throws BuilderException {
		try {
			return Long.parseLong(node.getArg(arg));
		} catch (IndexOutOfBoundsException e) {
			throw new BuilderException(String.format("No argument at position %d.", arg), node);
		} catch (NumberFormatException e) {
			throw new BuilderException(
				String.format("The string \"%s\" could not be parsed to a long int.", node.getArg(arg)), node);
		}
	}



	/**
	 * Takes a node and uses it as a key to search the scope for a node
	 * with a matching name and first argument.
	 * 
	 * @param node The node to use as a search key.
	 * @param scope The node whose children should be searched.
	 * @return The node with the same name, or {@code null} if none exist.
	 */
	public static @Nullable DataNode search(@NonNull DataNode node, @NonNull DataNode scope) {
		try {
			for (DataNode child : scope.getChildren()) {
				if (
					node.getName().equals(child.getName())
					&& node.getArg(0).equals(child.getArg(0))
				) {
					return child;
				}
			}
		} catch (IndexOutOfBoundsException e) {
			// This is not necessarily an error, so no error message is printed.
		}

		return null;
	}
}
